File: scripts/generate_voice.mjs
/**
 * VOICEVOX音声生成スクリプト
 * ずんだもん（ノーマル）で1.3倍速の音声を生成
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const VOICEVOX_URL = "http://127.0.0.1:50021";
const SPEAKER_ID = 3;  // ずんだもん（ノーマル）
const SPEED_SCALE = 1.3;  // 1.3倍速
const MAX_RETRIES = 3;
const RETRY_DELAY = 1000;  // 1秒

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * 台本ファイルを読み込んでセグメントに分割
 */
function readScript(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8').replace(/^\uFEFF/, '');

  // 行ごとに分割
  const lines = content.split(/\r?\n/);
  const segments = [];

  for (const line of lines) {
    const trimmed = line.trim();
    // 空行や記号だけの行はスキップ
    if (!trimmed) continue;
    // 番号だけの行やリストマーカーだけの行はスキップ
    if (/^[\s\*\-\d\.\:：・〜→]+$/.test(trimmed)) continue;
    // 先頭の記号やスペースを除去
    const cleaned = trimmed.replace(/^[\s\*\-・]+/, '').trim();
    if (cleaned) {
      segments.push(cleaned);
    }
  }

  return segments;
}

/**
 * VOICEVOXで音声を生成（リトライ機能付き）
 */
async function generateAudio(text, outputPath, speakerId = SPEAKER_ID, speed = SPEED_SCALE) {
  let lastError;

  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      // 音声合成用のクエリを作成
      const queryResponse = await fetch(
        `${VOICEVOX_URL}/audio_query?text=${encodeURIComponent(text)}&speaker=${speakerId}`,
        { method: 'POST' }
      );
      if (!queryResponse.ok) {
        throw new Error(`Query failed: ${queryResponse.status}`);
      }
      const query = await queryResponse.json();

      // 速度を設定
      query.speedScale = speed;

      // 音声合成
      const synthesisResponse = await fetch(
        `${VOICEVOX_URL}/synthesis?speaker=${speakerId}`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(query)
        }
      );
      if (!synthesisResponse.ok) {
        throw new Error(`Synthesis failed: ${synthesisResponse.status}`);
      }

      // ファイルに保存
      const buffer = Buffer.from(await synthesisResponse.arrayBuffer());
      fs.writeFileSync(outputPath, buffer);

      return outputPath;
    } catch (e) {
      lastError = e;
      if (attempt < MAX_RETRIES) {
        await sleep(RETRY_DELAY);
      }
    }
  }

  throw lastError;
}

async function main() {
  // パス設定
  const projectDir = path.resolve(__dirname, '..');
  const daihonPath = path.join(projectDir, 'src', 'assets', 'daihon.txt');
  const outputDir = path.join(projectDir, 'public', 'audio');

  // 出力ディレクトリ作成
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // 台本読み込み
  console.log(`台本を読み込み中: ${daihonPath}`);
  const segments = readScript(daihonPath);

  console.log(`セグメント数: ${segments.length}`);
  console.log(`話者: ずんだもん（ID: ${SPEAKER_ID}）`);
  console.log(`速度: ${SPEED_SCALE}倍速`);
  console.log('-'.repeat(50));

  // 各セグメントの音声を生成
  const generatedFiles = [];
  for (let i = 0; i < segments.length; i++) {
    const text = segments[i];
    const outputPath = path.join(outputDir, `segment_${String(i).padStart(3, '0')}.wav`);
    console.log(`[${i + 1}/${segments.length}] 生成中...`);
    console.log(`  テキスト: ${text.slice(0, 50)}${text.length > 50 ? '...' : ''}`);

    try {
      await generateAudio(text, outputPath);
      generatedFiles.push(outputPath);
      console.log(`  保存: segment_${String(i).padStart(3, '0')}.wav`);
    } catch (e) {
      console.log(`  エラー: ${e.message}`);
    }
  }

  console.log('-'.repeat(50));
  console.log(`完了！ ${generatedFiles.length}個の音声ファイルを生成しました`);
  console.log(`出力先: ${outputDir}`);

  // セグメント情報をJSONで保存
  const segmentsInfo = segments.map((text, i) => ({
    index: i,
    text: text,
    audio_file: `audio/segment_${String(i).padStart(3, '0')}.wav`
  }));

  const infoPath = path.join(outputDir, 'segments_info.json');
  fs.writeFileSync(infoPath, JSON.stringify(segmentsInfo, null, 2));
  console.log(`セグメント情報: ${infoPath}`);
}

main().catch(console.error);

File: scripts/generate_voice.py
#!/usr/bin/env python3
"""
VOICEVOX音声生成スクリプト
ずんだもん（ノーマル）で1.3倍速の音声を生成
"""

import requests
import json
import os
import re
from pathlib import Path

VOICEVOX_URL = "http://localhost:50021"
SPEAKER_ID = 3  # ずんだもん（ノーマル）
SPEED_SCALE = 1.3  # 1.3倍速

def read_script(file_path):
    """台本ファイルを読み込んでセグメントに分割"""
    with open(file_path, 'r', encoding='utf-8-sig') as f:
        content = f.read()

    # 空行で段落を分割し、空でないものだけ取得
    paragraphs = []
    for para in content.split('\n\n'):
        # 各段落の行を結合（改行を除去）
        lines = [line.strip() for line in para.strip().split('\n') if line.strip()]
        if lines:
            text = ''.join(lines)
            # 空白や記号だけの行はスキップ
            if re.sub(r'[\s\*\-\d\.\:：・]+', '', text):
                paragraphs.append(text)

    return paragraphs

def generate_audio(text, output_path, speaker_id=SPEAKER_ID, speed=SPEED_SCALE):
    """VOICEVOXで音声を生成"""
    # 音声合成用のクエリを作成
    query_response = requests.post(
        f"{VOICEVOX_URL}/audio_query",
        params={"text": text, "speaker": speaker_id}
    )
    query_response.raise_for_status()
    query = query_response.json()

    # 速度を設定
    query["speedScale"] = speed

    # 音声合成
    synthesis_response = requests.post(
        f"{VOICEVOX_URL}/synthesis",
        params={"speaker": speaker_id},
        json=query
    )
    synthesis_response.raise_for_status()

    # ファイルに保存
    with open(output_path, 'wb') as f:
        f.write(synthesis_response.content)

    return output_path

def main():
    # パス設定
    script_dir = Path(__file__).parent
    project_dir = script_dir.parent
    daihon_path = project_dir / "src" / "assets" / "daihon.txt"
    output_dir = project_dir / "public" / "audio"

    # 出力ディレクトリ作成
    output_dir.mkdir(parents=True, exist_ok=True)

    # 台本読み込み
    print(f"台本を読み込み中: {daihon_path}")
    segments = read_script(daihon_path)

    print(f"セグメント数: {len(segments)}")
    print(f"話者: ずんだもん（ID: {SPEAKER_ID}）")
    print(f"速度: {SPEED_SCALE}倍速")
    print("-" * 50)

    # 各セグメントの音声を生成
    generated_files = []
    for i, text in enumerate(segments):
        output_path = output_dir / f"segment_{i:03d}.wav"
        print(f"[{i+1}/{len(segments)}] 生成中...")
        print(f"  テキスト: {text[:50]}{'...' if len(text) > 50 else ''}")

        try:
            generate_audio(text, output_path)
            generated_files.append(str(output_path))
            print(f"  保存: {output_path.name}")
        except Exception as e:
            print(f"  エラー: {e}")

    print("-" * 50)
    print(f"完了！ {len(generated_files)}個の音声ファイルを生成しました")
    print(f"出力先: {output_dir}")

    # セグメント情報をJSONで保存
    segments_info = []
    for i, text in enumerate(segments):
        segments_info.append({
            "index": i,
            "text": text,
            "audio_file": f"audio/segment_{i:03d}.wav"
        })

    info_path = output_dir / "segments_info.json"
    with open(info_path, 'w', encoding='utf-8') as f:
        json.dump(segments_info, f, ensure_ascii=False, indent=2)
    print(f"セグメント情報: {info_path}")

if __name__ == "__main__":
    main()
